File: core/src/commonMain/kotlin/work/socialhub/kmastodon/internal/_UserStream.java
package work.socialhub.kmastodon.internal;

import com.google.gson.Gson;
import mastodon4j.MastodonException;
import mastodon4j.entity.Notification;
import mastodon4j.entity.Status;
import mastodon4j.streaming.LifeCycleListener;
import mastodon4j.streaming.UserStream;
import mastodon4j.streaming.UserStreamListener;
import net.socialhub.http.HttpRequestBuilder;
import net.socialhub.http.HttpResponse;
import net.socialhub.logger.Logger;

import java.io.BufferedReader;
import java.io.InputStreamReader;

/**
 * @author hecateball
 */
final class _UserStream implements UserStream {

    private static final Logger LOGGER = Logger.getLogger(_UserStream.class);

    private final HttpRequestBuilder builder;
    private LifeCycleListener lifeCycle;
    private _StreamEvent streamEvent;
    private Thread thread;
    private boolean isOpen = false;
    private final Gson gson;

    _UserStream(HttpRequestBuilder builder) {
        this.builder = builder;
        this.gson = _InternalUtility.getGsonInstance();
    }

    @Override
    public UserStream register(UserStreamListener listener, LifeCycleListener lifeCycle) {
        this.streamEvent = new _StreamEvent((event) -> {

            switch (event.getName()) {
                case "update":
                    listener.onUpdate(gson.fromJson(event.getData(), Status.class));
                    break;
                case "notification":
                    listener.onNotification(gson.fromJson(event.getData(), Notification.class));
                    break;
                case "delete":
                    listener.onDelete(gson.fromJson(event.getData(), Long.class));
                    break;
                default:
                    LOGGER.debug("Unexpected event name: " + event.getName());
            }
        });
        this.lifeCycle = lifeCycle;
        return this;
    }

    @Override
    public void open() {
        if (!isOpen) {
            thread = new Thread(() -> {
                try {
                    HttpResponse response = builder.get();
                    BufferedReader reader = new BufferedReader(new InputStreamReader(response.asStream(), "UTF-8"));
                    if (lifeCycle != null) {
                        lifeCycle.onConnect();
                    }
                    String line;

                    do {
                        line = reader.readLine();
                        this.streamEvent.add(line);
                        Thread.sleep(100);
                    } while (line != null);

                } catch (InterruptedException e) {
                    isOpen = false;
                    // close connection
                    if (lifeCycle != null) {
                        lifeCycle.onDisconnect();
                    }

                } catch (Exception e) {
                    isOpen = false;
                    // http exception
                    if (lifeCycle != null) {
                        lifeCycle.onDisconnect();
                    }
                    throw new MastodonException(e);
                }
            });
            thread.start();
            isOpen = true;
        }
    }

    @Override
    public void close() {
        if (isOpen) {
            thread.interrupt();
            isOpen = false;
        }
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/internal/_UserStream.kt
package work.socialhub.kmastodon.internal

import com.google.gson.Gson
import mastodon4j.MastodonException
import mastodon4j.entity.Notification
import mastodon4j.entity.Status
import mastodon4j.streaming.LifeCycleListener
import mastodon4j.streaming.UserStream
import mastodon4j.streaming.UserStreamListener
import net.socialhub.http.HttpRequestBuilder
import net.socialhub.http.HttpResponse
import net.socialhub.logger.Logger

/**
 * @author hecateball
 */
internal class _UserStream(builder: HttpRequestBuilder) : UserStream {
    private val builder: HttpRequestBuilder = builder
    private var lifeCycle: LifeCycleListener? = null
    private var streamEvent: _StreamEvent? = null
    private var thread: java.lang.Thread? = null
    private var isOpen = false
    private val gson: Gson = InternalUtility.getGsonInstance()

    fun register(listener: UserStreamListener, lifeCycle: LifeCycleListener?): UserStream {
        this.streamEvent = _StreamEvent { event ->
            when (event.getName()) {
                "update" -> listener.onUpdate(gson.fromJson(event.getData(), Status::class.java))
                "notification" -> listener.onNotification(gson.fromJson(event.getData(), Notification::class.java))
                "delete" -> listener.onDelete(gson.fromJson(event.getData(), Long::class.java))
                else -> LOGGER.debug("Unexpected event name: " + event.getName())
            }
        }
        this.lifeCycle = lifeCycle
        return this
    }

    fun open() {
        if (!isOpen) {
            thread = java.lang.Thread(java.lang.Runnable {
                try {
                    val response: HttpResponse = builder.get()
                    val reader: java.io.BufferedReader =
                        java.io.BufferedReader(java.io.InputStreamReader(response.asStream(), "UTF-8"))
                    if (lifeCycle != null) {
                        lifeCycle.onConnect()
                    }
                    var line: String

                    do {
                        line = reader.readLine()
                        streamEvent!!.add(line)
                        java.lang.Thread.sleep(100)
                    } while (line != null)
                } catch (e: java.lang.InterruptedException) {
                    isOpen = false
                    // close connection
                    if (lifeCycle != null) {
                        lifeCycle.onDisconnect()
                    }
                } catch (e: java.lang.Exception) {
                    isOpen = false
                    // http exception
                    if (lifeCycle != null) {
                        lifeCycle.onDisconnect()
                    }
                    throw MastodonException(e)
                }
            })
            thread.start()
            isOpen = true
        }
    }

    fun close() {
        if (isOpen) {
            thread.interrupt()
            isOpen = false
        }
    }

    companion object {
        private val LOGGER: Logger = Logger.getLogger(_UserStream::class.java)
    }
}
